olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,186 @@
+                download_stream = blob_client.download_blob()
+                download_file.write(download_stream.readall())
+
+        model_config.config["model_path"] = str(self.output_model_path / cloud_model_path)
+
+        return model_config
+
+    def get_hash_key(self, model_config: ModelConfig, pass_search_point: Dict[str, Any], input_model_hash: str):
+        hf_hub_model_commit_id = None
+        model_config_copy = deepcopy(model_config)
+        if input_model_hash is None:
+            from huggingface_hub import repo_info
+
+            if model_config.has_hf_config():
+                hf_hub_model_commit_id = repo_info(model_config.get_hf_model_name()).sha
+        else:
+            model_config_copy.config.pop("model_path", None)
+        return hash_dict(
+            {
+                "hf_hub_model_commit_id": hf_hub_model_commit_id,
+                "model_config": model_config_copy.to_json(),
+                "pass_search_point": pass_search_point,
+                "input_model_hash": input_model_hash,
+            }
+        )
+
+    def exist_in_cloud_cache(self, output_model_hash: str):
+        logger.info("Checking cloud cache for model hash: %s", output_model_hash)
+        return self.container_client.get_blob_client(output_model_hash).exists()
+
+    def get_model_config_by_hash_key(self, output_model_hash: str):
+        model_config_blob = f"{output_model_hash}/model_config.json"
+        model_config_path = Path(self.cache_dir) / output_model_hash
+        model_config_path.mkdir(parents=True, exist_ok=True)
+        model_config_path = model_config_path / "model_config.json"
+
+        blob_client = self.container_client.get_blob_client(model_config_blob)
+        logger.info("Downloading %s to %s", model_config_blob, model_config_path)
+
+        with open(model_config_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
+            download_file.write(download_stream.readall())
+
+        # model config
+        with open(model_config_path) as file:
+            model_config_dict = json.load(file)
+            return ModelConfig.from_json(model_config_dict)
+
+    def upload_model_to_cloud_cache(self, output_model_hash: str, output_model_config: ModelConfig):
+        logger.info("Uploading model to cloud cache.")
+        model_path = output_model_config.config.get("model_path")
+
+        if model_path is None:
+            logger.error("Model path is not found in the output model config. Upload failed.")
+            return
+
+        model_path = Path(model_path)
+        model_config_copy = deepcopy(output_model_config)
+        model_config_copy.config["model_path"] = None
+
+        # upload model config file
+        model_config_bytes = json.dumps(model_config_copy.to_json()).encode()
+        with io.BytesIO(model_config_bytes) as data:
+            self.container_client.upload_blob(f"{output_model_hash}/model_config.json", data=data, overwrite=False)
```

Comment:
You might want to move this after all the files/folders are uploaded to avoid scenarios where the `model_config.json` is available in the cache but the necessary files are still uploading.

Reply:
Yeah we probably should upload large files first. Updated.